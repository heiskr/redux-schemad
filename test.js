const {
  createReducer,
  combineReducers,
} = require('@acemarke/redux-starter-kit')
const {
  mapObj,
  // setField,
  // updateField,
  resetField,
  addItem,
  // addItems,
  updateItem,
  // updateItems,
  removeItem,
  // removeItems,
  // resetItems,
  // createAction,
  // createFieldActionsMap,
  // createGlobalActionsMap,
  // createCollectionActionsMap,
  createDefaultState,
  createSchemadReducer,
  createActionTypes,
  createActions,
  verifyStateKeys,
  verifyStateKeysMiddleware,
} = require('./index')

describe('Redux Schemad', () => {
  const schema = {
    global: {
      loggedInUserId: 'erty1234',
      settings: {},
    },
    users: {
      '[id]': {
        id: '',
        name: 'tester',
      },
    },
  }
  const defaultState = createDefaultState(schema)
  const reducer = combineReducers(createSchemadReducer(schema, createReducer))
  const actions = createActions(schema)
  const actionTypes = createActionTypes(schema)

  describe('1 Actions', () => {
    test('match snapshot', () => {
      expect(Object.keys(actions)).toMatchSnapshot()
    })
  })

  describe('2 Action Types', () => {
    test('match snapshot', () => {
      expect(actionTypes).toMatchSnapshot()
    })
  })

  describe('3 Default State', () => {
    test('match snapshot', () => {
      expect(defaultState).toMatchSnapshot()
    })
  })

  describe('4 Reducer', () => {
    describe('1 Field', () => {
      test('1 set a field', () => {
        const action = actions.setLoggedInUserId('abcd1234')
        expect(action).toMatchSnapshot()
        expect(reducer(defaultState, action)).toMatchSnapshot()
      })

      test('2 update a flat field', () => {
        const action = actions.updateLoggedInUserId('abcd1234')
        expect(action).toMatchSnapshot()
        expect(reducer(defaultState, action)).toMatchSnapshot()
      })

      test('3 update an object field', () => {
        const action = actions.updateSettings({ email: false })
        expect(action).toMatchSnapshot()
        expect(reducer(defaultState, action)).toMatchSnapshot()
      })

      test('4 update an object field not before though', () => {
        const state = {
          ...defaultState,
          global: { loggedInUserId: 'erty1234', settings: null },
        }
        const action = actions.updateSettings({ email: false })
        expect(action).toMatchSnapshot()
        expect(reducer(state, action)).toMatchSnapshot()
      })

      test('5 reset a field', () => {
        const state = {
          ...defaultState,
          global: { loggedInUserId: 'abcd1234', settings: {} },
        }
        const action = actions.resetLoggedInUserId()
        expect(action).toMatchSnapshot()
        expect(reducer(state, action)).toMatchSnapshot()
      })
    })

    describe('2 Collection', () => {
      test('1 add a child', () => {
        const action = actions.addUser({
          id: 'abcd1234',
        })
        expect(action).toMatchSnapshot()
        expect(reducer(defaultState, action)).toMatchSnapshot()
      })

      test('2 add many children', () => {
        const action = actions.addUsers([
          {
            id: 'abcd1234',
          },
          {
            id: 'wxyz0987',
          },
        ])
        expect(action).toMatchSnapshot()
        expect(reducer(defaultState, action)).toMatchSnapshot()
      })

      test('3 update a child', () => {
        const state = {
          ...defaultState,
          users: { abcd1234: {} },
        }
        const action = actions.updateUser({
          id: 'abcd1234',
          name: 'foobar',
        })
        expect(action).toMatchSnapshot()
        expect(reducer(state, action)).toMatchSnapshot()
      })

      test('4 update many children', () => {
        const state = {
          ...defaultState,
          users: { abcd1234: {} },
        }
        const action = actions.updateUsers([
          {
            id: 'abcd1234',
            name: 'foobar',
          },
          {
            id: 'wxyz0987',
          },
        ])
        expect(action).toMatchSnapshot()
        expect(reducer(state, action)).toMatchSnapshot()
      })

      test('5 remove a child', () => {
        const state = {
          ...defaultState,
          users: { abcd1234: { id: 'abcd1234' }, wxyz0987: { id: 'wxyz0987' } },
        }
        const action = actions.removeUser({
          id: 'abcd1234',
        })
        expect(action).toMatchSnapshot()
        expect(reducer(state, action)).toMatchSnapshot()
      })

      test('6 remove many children', () => {
        const state = {
          ...defaultState,
          users: { abcd1234: { id: 'abcd1234' }, wxyz0987: { id: 'wxyz0987' } },
        }
        const action = actions.removeUsers([
          {
            id: 'abcd1234',
          },
        ])
        expect(action).toMatchSnapshot()
        expect(reducer(state, action)).toMatchSnapshot()
      })

      test('7 reset the collection', () => {
        const state = {
          ...defaultState,
          users: { abcd1234: { id: 'abcd1234' } },
        }
        const action = actions.resetUsers()
        expect(action).toMatchSnapshot()
        expect(reducer(state, action)).toMatchSnapshot()
      })
    })
  })

  describe('5 Verify State', () => {
    const goodState = {
      global: {
        loggedInUserId: 'erty1234',
        settings: {},
      },
      users: {},
    }

    const badState = {
      global: {
        loggedInUserId: 'erty1234',
      },
      users: {
        a: {
          name: 'abcd',
          email: 'abcd@example.com',
        },
      },
      foo: {},
    }

    describe('1 verifyStateKeys', () => {
      test('1 it should not find any errors', () => {
        expect(verifyStateKeys(schema, goodState)).toHaveLength(0)
      })

      test('2 it should find all errors', () => {
        expect(verifyStateKeys(schema, badState)).toHaveLength(3)
      })
    })

    describe('2 verifyStateKeysMiddleware', () => {
      /* eslint-disable no-console */
      const prevErr = console.error

      beforeEach(() => {
        console.error = jest.fn()
      })

      afterEach(() => {
        console.error = prevErr
      })

      test('1 should not report errors', () => {
        const store = { getState: () => goodState }
        const next = jest.fn()
        const action = {}
        verifyStateKeysMiddleware(schema)(store)(next)(action)
        expect(console.error).not.toBeCalled()
      })

      test('2 should report errors', () => {
        const store = { getState: () => badState }
        const next = jest.fn()
        const action = {}
        verifyStateKeysMiddleware(schema)(store)(next)(action)
        expect(console.error).toMatchSnapshot()
      })
      /* eslint-enable */
    })
  })

  describe('6 Extras', () => {
    test('1 call other things...', () => {
      mapObj({})
      resetField()
      addItem()
      updateItem()
      removeItem()
    })
  })
})
