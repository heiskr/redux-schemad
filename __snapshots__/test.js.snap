// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redux Schemad 1 Actions match snapshot 1`] = `
Array [
  "setLoggedInUserId",
  "updateLoggedInUserId",
  "resetLoggedInUserId",
  "setSettings",
  "updateSettings",
  "resetSettings",
  "addUser",
  "addUsers",
  "updateUser",
  "updateUsers",
  "removeUser",
  "removeUsers",
  "resetUsers",
]
`;

exports[`Redux Schemad 2 Action Types match snapshot 1`] = `
Object {
  "ADD_USER": "ADD_USER",
  "ADD_USERS": "ADD_USERS",
  "REMOVE_USER": "REMOVE_USER",
  "REMOVE_USERS": "REMOVE_USERS",
  "RESET_LOGGED_IN_USER_ID": "RESET_LOGGED_IN_USER_ID",
  "RESET_SETTINGS": "RESET_SETTINGS",
  "RESET_USERS": "RESET_USERS",
  "SET_LOGGED_IN_USER_ID": "SET_LOGGED_IN_USER_ID",
  "SET_SETTINGS": "SET_SETTINGS",
  "UPDATE_LOGGED_IN_USER_ID": "UPDATE_LOGGED_IN_USER_ID",
  "UPDATE_SETTINGS": "UPDATE_SETTINGS",
  "UPDATE_USER": "UPDATE_USER",
  "UPDATE_USERS": "UPDATE_USERS",
}
`;

exports[`Redux Schemad 3 Default State match snapshot 1`] = `
Object {
  "global": Object {
    "loggedInUserId": "erty1234",
    "settings": Object {},
  },
  "users": Object {},
}
`;

exports[`Redux Schemad 4 Reducer 1 Field 1 set a field 1`] = `
Object {
  "meta": undefined,
  "payload": "abcd1234",
  "type": "SET_LOGGED_IN_USER_ID",
}
`;

exports[`Redux Schemad 4 Reducer 1 Field 1 set a field 2`] = `
Object {
  "global": Object {
    "loggedInUserId": "abcd1234",
    "settings": Object {},
  },
  "users": Object {},
}
`;

exports[`Redux Schemad 4 Reducer 1 Field 2 update a flat field 1`] = `
Object {
  "meta": undefined,
  "payload": "abcd1234",
  "type": "UPDATE_LOGGED_IN_USER_ID",
}
`;

exports[`Redux Schemad 4 Reducer 1 Field 2 update a flat field 2`] = `
Object {
  "global": Object {
    "loggedInUserId": "abcd1234",
    "settings": Object {},
  },
  "users": Object {},
}
`;

exports[`Redux Schemad 4 Reducer 1 Field 3 update an object field 1`] = `
Object {
  "meta": undefined,
  "payload": Object {
    "email": false,
  },
  "type": "UPDATE_SETTINGS",
}
`;

exports[`Redux Schemad 4 Reducer 1 Field 3 update an object field 2`] = `
Object {
  "global": Object {
    "loggedInUserId": "erty1234",
    "settings": Object {
      "email": false,
    },
  },
  "users": Object {},
}
`;

exports[`Redux Schemad 4 Reducer 1 Field 4 update an object field not before though 1`] = `
Object {
  "meta": undefined,
  "payload": Object {
    "email": false,
  },
  "type": "UPDATE_SETTINGS",
}
`;

exports[`Redux Schemad 4 Reducer 1 Field 4 update an object field not before though 2`] = `
Object {
  "global": Object {
    "loggedInUserId": "erty1234",
    "settings": Object {
      "email": false,
    },
  },
  "users": Object {},
}
`;

exports[`Redux Schemad 4 Reducer 1 Field 5 reset a field 1`] = `
Object {
  "meta": undefined,
  "payload": undefined,
  "type": "RESET_LOGGED_IN_USER_ID",
}
`;

exports[`Redux Schemad 4 Reducer 1 Field 5 reset a field 2`] = `
Object {
  "global": Object {
    "loggedInUserId": "erty1234",
    "settings": Object {},
  },
  "users": Object {},
}
`;

exports[`Redux Schemad 4 Reducer 2 Collection 1 add a child 1`] = `
Object {
  "meta": undefined,
  "payload": Object {
    "id": "abcd1234",
  },
  "type": "ADD_USER",
}
`;

exports[`Redux Schemad 4 Reducer 2 Collection 1 add a child 2`] = `
Object {
  "global": Object {
    "loggedInUserId": "erty1234",
    "settings": Object {},
  },
  "users": Object {
    "abcd1234": Object {},
  },
}
`;

exports[`Redux Schemad 4 Reducer 2 Collection 2 add many children 1`] = `
Object {
  "meta": undefined,
  "payload": Array [
    Object {
      "id": "abcd1234",
    },
    Object {
      "id": "wxyz0987",
    },
  ],
  "type": "ADD_USERS",
}
`;

exports[`Redux Schemad 4 Reducer 2 Collection 2 add many children 2`] = `
Object {
  "global": Object {
    "loggedInUserId": "erty1234",
    "settings": Object {},
  },
  "users": Object {
    "abcd1234": Object {},
    "wxyz0987": Object {},
  },
}
`;

exports[`Redux Schemad 4 Reducer 2 Collection 3 update a child 1`] = `
Object {
  "meta": undefined,
  "payload": Object {
    "id": "abcd1234",
    "name": "foobar",
  },
  "type": "UPDATE_USER",
}
`;

exports[`Redux Schemad 4 Reducer 2 Collection 3 update a child 2`] = `
Object {
  "global": Object {
    "loggedInUserId": "erty1234",
    "settings": Object {},
  },
  "users": Object {
    "abcd1234": Object {
      "name": "foobar",
    },
  },
}
`;

exports[`Redux Schemad 4 Reducer 2 Collection 4 update many children 1`] = `
Object {
  "meta": undefined,
  "payload": Array [
    Object {
      "id": "abcd1234",
      "name": "foobar",
    },
    Object {
      "id": "wxyz0987",
    },
  ],
  "type": "UPDATE_USERS",
}
`;

exports[`Redux Schemad 4 Reducer 2 Collection 4 update many children 2`] = `
Object {
  "global": Object {
    "loggedInUserId": "erty1234",
    "settings": Object {},
  },
  "users": Object {
    "abcd1234": Object {
      "name": "foobar",
    },
    "wxyz0987": Object {},
  },
}
`;

exports[`Redux Schemad 4 Reducer 2 Collection 5 remove a child 1`] = `
Object {
  "meta": undefined,
  "payload": Object {
    "id": "abcd1234",
  },
  "type": "REMOVE_USER",
}
`;

exports[`Redux Schemad 4 Reducer 2 Collection 5 remove a child 2`] = `
Object {
  "global": Object {
    "loggedInUserId": "erty1234",
    "settings": Object {},
  },
  "users": Object {
    "wxyz0987": Object {
      "id": "wxyz0987",
    },
  },
}
`;

exports[`Redux Schemad 4 Reducer 2 Collection 6 remove many children 1`] = `
Object {
  "meta": undefined,
  "payload": Array [
    Object {
      "id": "abcd1234",
    },
  ],
  "type": "REMOVE_USERS",
}
`;

exports[`Redux Schemad 4 Reducer 2 Collection 6 remove many children 2`] = `
Object {
  "global": Object {
    "loggedInUserId": "erty1234",
    "settings": Object {},
  },
  "users": Object {
    "wxyz0987": Object {
      "id": "wxyz0987",
    },
  },
}
`;

exports[`Redux Schemad 4 Reducer 2 Collection 7 reset the collection 1`] = `
Object {
  "meta": undefined,
  "payload": undefined,
  "type": "RESET_USERS",
}
`;

exports[`Redux Schemad 4 Reducer 2 Collection 7 reset the collection 2`] = `
Object {
  "global": Object {
    "loggedInUserId": "erty1234",
    "settings": Object {},
  },
  "users": Object {},
}
`;

exports[`Redux Schemad 5 Verify State 2 verifyStateKeysMiddleware 2 should report errors 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Top state has extra keys: foo
\\"global\\" schema has extra keys: settings
\\"users.a\\" state has extra keys: email",
    ],
  ],
}
`;
